server/service/src/rnr_form/generate_rnr_form_lines.rs

```diff
@@ -1,95 +1,339 @@
+                0.0
+            };
 
-            let amc = item_stat.average_monthly_consumption;
-            let maximum_quantity = amc * target_mos as f64;
+            let earliest_expiry = get_earliest_expiry(&ctx.connection, store_id, &item_id)?;
 
-            RnRFormLineRow {
+            Ok(RnRFormLineRow {
                 id: uuid(),
                 rnr_form_id: rnr_form_id.to_string(),
-                item_id: item_stat.item_id,
+                item_id,
                 average_monthly_consumption: amc,
-                // from prev rnr, if exists, else bal on opening date
-                initial_balance: 0.0,
-
-                // incoming over period
-                quantity_received,
-                quantity_consumed,
+                initial_balance,
+                quantity_received: usage.replenished,
+                quantity_consumed: usage.consumed,
                 stock_out_duration,
-                // all adjustments (adjustments and returns over period)
-                adjustments: 0.0,
+                adjustments: usage.adjusted,
 
                 adjusted_quantity_consumed,
-                // SOH on date
                 final_balance,
                 maximum_quantity,
-                // stock lines for item, find earliest expiry or blank
-                expiry_date: None,
-                requested_quantity: maximum_quantity - final_balance,
+                expiry_date: earliest_expiry,
+                requested_quantity,
                 comment: None,
                 confirmed: false,
-            }
+            })
         })
-        .collect();
+        .collect::<Result<Vec<RnRFormLineRow>, RepositoryError>>();
+
+    rnr_form_lines
+}
+
+fn get_master_list_item_ids(
+    ctx: &ServiceContext,
+    master_list_id: &str,
+) -> Result<Vec<String>, RepositoryError> {
+    MasterListLineRepository::new(&ctx.connection)
+        .query_by_filter(
+            MasterListLineFilter::new().master_list_id(EqualFilter::equal_to(master_list_id)),
+        )
+        .map(|lines| lines.into_iter().map(|line| line.item_id).collect())
+}
+
+fn get_last_rnr_form_lines(
+    connection: &StorageConnection,
+    previous_form_id: Option<String>,
+) -> Result<HashMap<String, RnRFormLineRow>, RepositoryError> {
+    let mut form_lines_by_item_id = HashMap::new();
+
+    if let Some(previous_form_id) = previous_form_id {
+        let rows = RnRFormLineRowRepository::new(connection)
+            .find_many_by_rnr_form_id(&previous_form_id)?;
+
+        for row in rows.into_iter() {
+            form_lines_by_item_id.insert(row.item_id.clone(), row);
+        }
+    }
+
+    Ok(form_lines_by_item_id)
+}
+
+pub fn get_opening_balance(
+    connection: &StorageConnection,
+    previous_row: Option<&RnRFormLineRow>,
+    store_id: &str,
+    item_id: &str,
+    start_date: NaiveDate,
+) -> Result<f64, RepositoryError> {
+    if let Some(previous_row) = previous_row {
+        return Ok(previous_row.final_balance);
```

Comment:
There is a risk here, if there is a gap between previous R&R form period end date, and this period start date...

Is that worth checking here?